File: src-tauri/src/commands/memo.rs
use crate::database::get_conn;
use crate::models::memo::MemoDetail;
use crate::models::MemoIndexItem;
use crate::repositories::{MemoRepository, WorkspaceRepository};
use serde::Deserialize;
use tauri::command;

#[derive(Deserialize)]
pub struct GetMemosArgs {
    pub workspace_slug_name: String,
}

#[command]
pub fn get_workspace_memos(args: GetMemosArgs) -> Result<Vec<MemoIndexItem>, String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    MemoRepository::list(&conn, workspace.id).map_err(|e| e.to_string())
}

#[derive(Deserialize)]
pub struct GetMemoArgs {
    pub workspace_slug_name: String,
    pub memo_slug_title: String,
}

#[command]
pub fn get_memo(args: GetMemoArgs) -> Result<MemoDetail, String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    match MemoRepository::find_by_slug(&conn, workspace.id, &args.memo_slug_title) {
        Ok(Some(memo)) => Ok(memo),
        Ok(None) => Err(format!("Memo not found for slug: {}", args.memo_slug_title)),
        Err(e) => Err(e.to_string()),
    }
}

#[derive(Deserialize)]
pub struct CreateMemoArgs {
    pub workspace_slug_name: String,
    pub slug_title: String,
    pub title: String,
    pub content: String,
}

#[command]
pub fn create_memo(args: CreateMemoArgs) -> Result<MemoDetail, String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| {
            format!(
                "Workspace not found for slug: {}",
                &args.workspace_slug_name
            )
        })?;

    MemoRepository::create(
        &conn,
        workspace.id,
        &args.slug_title,
        &args.title,
        &args.content,
    )
    .map_err(|e| e.to_string())
}

#[derive(Deserialize)]
pub struct SaveMemoArgs {
    pub workspace_slug_name: String,
    pub target_slug_title: String,
    pub new_slug_title: String,
    pub new_title: String,
    pub new_content: String,
    pub new_description: String,
}

#[command]
pub fn save_memo(args: SaveMemoArgs) -> Result<(), String> {
    let mut conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| {
            format!(
                "Workspace not found for slug: {}",
                &args.workspace_slug_name
            )
        })?;

    MemoRepository::save(
        &mut conn,
        workspace.id,
        &args.workspace_slug_name,
        &args.target_slug_title,
        &args.new_slug_title,
        &args.new_title,
        &args.new_content,
        &args.new_description,
    )
    .map_err(|e| e.to_string())
}

#[derive(Deserialize)]
pub struct DeleteMemoArgs {
    pub workspace_slug_name: String,
    pub memo_slug_title: String,
}

#[command]
pub fn delete_memo(args: DeleteMemoArgs) -> Result<(), String> {
    let mut conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| {
            format!(
                "Workspace not found for slug: {}",
                &args.workspace_slug_name
            )
        })?;

    let memo = MemoRepository::find_by_slug(&conn, workspace.id, &args.memo_slug_title)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Memo not found for slug: {}", args.memo_slug_title))?;

    MemoRepository::delete(&mut conn, memo.id).map_err(|e| e.to_string())?;

    Ok(())
}

File: src/external/tauri/memo.ts
import { invokeCommand } from './invoker';

import type { MemoDetail, MemoIndexItem } from '~/models/memo';

import { encodeForSlug } from '~/utils/slug';

export const memoCommand = {
  list: async (workspace: { slugName: string }) => {
    return await invokeCommand<MemoIndexItem[]>('get_workspace_memos', {
      workspace_slug_name: workspace.slugName,
    });
  },

  create: async (memo: { workspaceSlugName: string; title: string }) => {
    return await invokeCommand<MemoDetail>('create_memo', {
      workspace_slug_name: memo.workspaceSlugName,
      slug_title: encodeForSlug(memo.title),
      title: memo.title,
      content: JSON.stringify(''),
    });
  },

  get: async (memo: { workspaceSlugName: string; memoSlugTitle: string }) => {
    return await invokeCommand<MemoDetail>('get_memo', {
      workspace_slug_name: memo.workspaceSlugName,
      memo_slug_title: memo.memoSlugTitle,
    });
  },

  save: async (
    memo: { workspaceSlug: string; memoSlug: string },
    newMemo: {
      slugTitle: string;
      title: string;
      content: string;
      description: string;
      thumbnailImage?: string;
    },
  ) => {
    await invokeCommand('save_memo', {
      workspace_slug_name: memo.workspaceSlug,
      target_slug_title: memo.memoSlug,
      new_slug_title: newMemo.slugTitle,
      new_title: newMemo.title,
      new_content: newMemo.content,
      new_description: newMemo.description,
      new_thumbnail_image: newMemo.thumbnailImage || '',
    });
  },

  trash: async (memo: { workspaceSlug: string; memoSlug: string }) => {
    await invokeCommand('delete_memo', {
      workspace_slug_name: memo.workspaceSlug,
      memo_slug_title: memo.memoSlug,
    });
  },
};
